### PR TITLE
Board Editor Layout

### PR DESCRIPTION
**Pull request by SJuliez**
Makes the BoardEditor adapt to the gui scaling. Also some code cleanup.
Also moves the reaction to gui scale changes (writing them to the GUI Prefs and creating a Prefchange event) via menu or keyboard shortcut from the ClientGUI (which only operates in lobby and game) to the commonmenubar itself which operates everywhere.

### CODE REVIEW

File: megamek/src/megamek/client/ui/swing/CommonMenuBar.java [hunk 1]
```diff
@@ -826,6 +826,20 @@ private JMenuItem createMenuItem(JMenu m, String label, String command) {
...
+            float guiScale = GUIPreferences.getInstance().getGUIScale();
+            if (guiScale < ClientGUI.MAX_GUISCALE) {
+                GUIPreferences.getInstance().setValue(GUIPreferences.GUI_SCALE, guiScale + 0.1);
+            }
+        } else if (event.getActionCommand().equals(ClientGUI.VIEW_DECGUISCALE)) {
+            float guiScale = GUIPreferences.getInstance().getGUIScale();
+            if (guiScale > ClientGUI.MIN_GUISCALE) {
+                GUIPreferences.getInstance().setValue(GUIPreferences.GUI_SCALE, guiScale - 0.1);
+            }
+        }
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/CommonMenuBar.java:841]: Should we return at the end of these two, even if the if statement doesn't occur?

Comment (SJuliez) [reply megamek/src/megamek/client/ui/swing/CommonMenuBar.java:841]: You confuse me :) I thought about returning when the action command is in fact one of the GUISCALE things. But I thought well why not pass it on, if for whatever reason the other components want to do something with it as well. So, no return there, the event is always passed on to e.g. the ClientGUI, even if its a GUISCALE event.

Comment (Windchild292) [reply megamek/src/megamek/client/ui/swing/CommonMenuBar.java:841]: I missed the word internal, my apologies... but that makes a lot of sense :)

File: megamek/src/megamek/client/ui/swing/BoardEditor.java [hunk 2]
```diff
@@ -299,9 +279,19 @@ public void setTerrainTypes(List<TerrainTypeHelper> terrainTypes) {
...
     private Coords lastClicked;
+    private JLabel labTheme = new JLabel(Messages.getString("BoardEditor.labTheme"), SwingConstants.LEFT);
+    
+    private FixedYPanel panelHexSettings = new FixedYPanel();
+    private FixedYPanel panelTerrSettings = new FixedYPanel(new GridLayout(0, 2, 4, 4));
+    private FixedYPanel panelBoardSettings = new FixedYPanel();
+    
+    // Help Texts
+    JLabel labHelp1 = new JLabel(Messages.getString("BoardEditor.helpText"),SwingConstants.LEFT);
+    JLabel labHelp2 = new JLabel(Messages.getString("BoardEditor.helpText2"),SwingConstants.LEFT);
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/BoardEditor.java:290]:
```suggestion
    private JLabel labHelp1 = new JLabel(Messages.getString("BoardEditor.helpText"), SwingConstants.LEFT);
    private JLabel labHelp2 = new JLabel(Messages.getString("BoardEditor.helpText2"), SwingConstants.LEFT);
```

File: megamek/src/megamek/client/ui/swing/BoardEditor.java [hunk 3]
```diff
@@ -544,37 +530,36 @@ public void windowClosing(WindowEvent e) {
...
 
     /**
-     * Sets up JButtons
+     * Sets up Scaling Icon Buttons
      */
-    private JButton prepareButton(String iconName, String buttonName, ArrayList<JButton> bList) {
-        JButton button = new JButton(buttonName);
-        button.addActionListener(this);
+    private ScalingIconButton prepareButton(String iconName, String buttonName, 
+            ArrayList<ScalingIconButton> bList, int width) {
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/BoardEditor.java:545]:
```suggestion
    private @Nullable ScalingIconButton prepareButton(String iconName, String buttonName, 
            List<ScalingIconButton> bList, int width) {
```

File: megamek/src/megamek/client/ui/swing/BoardEditor.java [hunk 4]
```diff
@@ -584,92 +569,84 @@ private JButton prepareButton(String iconName, String buttonName, ArrayList<JBut
...
     /**
-     * Sets up JToggleButtons
+     * Sets up Scaling Icon ToggleButtons
      */
-    private JToggleButton addTerrainTButton(String iconName, String buttonName, ArrayList<JToggleButton> bList) {
-        JToggleButton button = new JToggleButton(buttonName);
-        button.addActionListener(this);
-        
+    private ScalingIconToggleButton prepareToggleButton(String iconName, String buttonName, 
+            ArrayList<ScalingIconToggleButton> bList, int width) {
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/BoardEditor.java:580]:
```suggestion
    private @Nullable ScalingIconToggleButton prepareToggleButton(String iconName, String buttonName, 
            List<ScalingIconToggleButton> bList, int width) {
```

File: megamek/src/megamek/client/ui/swing/BoardEditor.java [hunk 5]
```diff
@@ -2164,6 +2032,58 @@ public void valueChanged(ListSelectionEvent event) {
...
+        texElev.setFont(scaledFont);
+        texTerrainLevel.setFont(scaledFont);
+        
+        labHelp1.setFont(scaledFont);
+        labHelp2.setFont(scaledFont);
+        labTheme.setFont(scaledFont);
+        
+        ((TitledBorder)panelBoardSettings.getBorder()).setTitleFont(scaledFont);
+        ((TitledBorder)panelHexSettings.getBorder()).setTitleFont(scaledFont);
+        ((TitledBorder)panelTerrSettings.getBorder()).setTitleFont(scaledFont);
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/BoardEditor.java:2073]:
```suggestion
        ((TitledBorder) panelBoardSettings.getBorder()).setTitleFont(scaledFont);
        ((TitledBorder) panelHexSettings.getBorder()).setTitleFont(scaledFont);
        ((TitledBorder) panelTerrSettings.getBorder()).setTitleFont(scaledFont);
```

File: megamek/src/megamek/client/ui/swing/BoardEditor.java [hunk 6]
```diff
@@ -2378,4 +2300,95 @@ public int getNumber() {
...
+                setDisabledIcon(new ImageIcon(ImageUtil.getScaledImage(baseDisabledImage, realWidth, realHeight)));
+            }
+        }
+        
+        void setRolloverImage(Image image) {
+            baseRolloverImage = image;
+        }
+        
+        void setDisabledImage(Image image) {
+            baseDisabledImage = image;
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/BoardEditor.java:2346]:
```suggestion
        void setRolloverImage(@Nullable Image image) {
            baseRolloverImage = image;
        }
        
        void setDisabledImage(@Nullable Image image) {
            baseDisabledImage = image;
```

File: megamek/src/megamek/client/ui/swing/BoardEditor.java [hunk 7]
```diff
@@ -2378,4 +2300,95 @@ public int getNumber() {
...
+                realHeight = baseSelectedImage.getHeight(null) * realWidth / baseSelectedImage.getWidth(null);
+                setSelectedIcon(new ImageIcon(ImageUtil.getScaledImage(baseSelectedImage, realWidth, realHeight)));
+            }
+        }
+        
+        void setRolloverImage(Image image) {
+            baseRolloverImage = image;
+        }
+        
+        void setSelectedImage(Image image) {
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/BoardEditor.java:2390]:
```suggestion
        void setRolloverImage(@Nullable Image image) {
            baseRolloverImage = image;
        }
        
        void setSelectedImage(@Nullable Image image) {
```

**Review by Windchild292 — COMMENTED** [review]
Looks good, just a few comments first. Also, can we try using @Nullable whenever there is a potential null return, as that should help prevent some of the NPEs we get in the future.

**Comment by SJuliez** [comment]
@Windchild292 changes are in; I changed a few things concerning the @Nullable: The prepareButton methods shouldnt return null and I changed the ScalingXXButtons to actually work correctly when passing a null image. I opted for a comment on passing null rather than the annotation as the comment is more helpful. I didnt fetch the changes I committed directly here (this always throws me) so they may show up again in the latest commit.

File: megamek/src/megamek/client/ui/swing/BoardEditor.java [hunk 8]
```diff
@@ -299,9 +279,19 @@ public void setTerrainTypes(List<TerrainTypeHelper> terrainTypes) {
...
     private Coords lastClicked;
+    private JLabel labTheme = new JLabel(Messages.getString("BoardEditor.labTheme"), SwingConstants.LEFT);
+    
+    private FixedYPanel panelHexSettings = new FixedYPanel();
+    private FixedYPanel panelTerrSettings = new FixedYPanel(new GridLayout(0, 2, 4, 4));
+    private FixedYPanel panelBoardSettings = new FixedYPanel();
+    
+    // Help Texts
+    private JLabel labHelp1 = new JLabel(Messages.getString("BoardEditor.helpText"),SwingConstants.LEFT);
+    private JLabel labHelp2 = new JLabel(Messages.getString("BoardEditor.helpText2"),SwingConstants.LEFT);
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/BoardEditor.java:290]:
```suggestion
    private JLabel labHelp1 = new JLabel(Messages.getString("BoardEditor.helpText"), SwingConstants.LEFT);
    private JLabel labHelp2 = new JLabel(Messages.getString("BoardEditor.helpText2"), SwingConstants.LEFT);
```

**Review by Windchild292 — COMMENTED** [review]
One last piece, otherwise it looks good. I'd still like the nullable annotation just for a quick glance, but the comment is good :)

**Pull request merged**